nameko/testing/services.py

```diff
@@ -48,93 +48,157 @@ def entrypoint_hook(container, method_name, context_data=None):
                 method_name, container))
 
     def hook(*args, **kwargs):
-        result = event.Event()
-
-        def handle_result(worker_ctx, res=None, exc_info=None):
-            result.send(res, exc_info)
-            return res, exc_info
-
-        container.spawn_worker(entrypoint, args, kwargs,
-                               context_data=context_data,
-                               handle_result=handle_result)
+        hook_result = event.Event()
+
+        def wait_for_entrypoint():
+            with entrypoint_waiter(container, method_name) as waiter_result:
+                container.spawn_worker(
+                    entrypoint, args, kwargs,
+                    context_data=context_data
+                )
+            try:
+                hook_result.send(waiter_result.get())
+            except Exception as exc:
+                hook_result.send_exception(exc)
 
-        # If the container errors (e.g. due to a bad entrypoint), handle_result
-        # is never called and we hang. To mitigate, we spawn a greenlet waiting
-        # for the container, and if that throws we send the exception back
-        # as our result
-        def catch_container_errors(gt):
+        def wait_for_container():
             try:
-                gt.wait()
+                container.wait()
             except Exception as exc:
-                result.send_exception(exc)
+                if not hook_result.ready():
+                    hook_result.send_exception(exc)
 
-        gt = eventlet.spawn(container.wait)
-        gt.link(catch_container_errors)
+        # If the container errors (e.g. due to a bad entrypoint), the
+        # entrypoint_waiter never completes. To mitigate, we exit the hook
+        # on the first greenthread to complete.
+        eventlet.spawn_n(wait_for_entrypoint)
+        eventlet.spawn_n(wait_for_container)
 
-        return result.wait()
+        return hook_result.wait()
 
     yield hook
 
 
-class EntrypointWaiter(DependencyProvider):
-    """Helper for `entrypoint_waiter`
+class EntrypointWaiterTimeout(Exception):
```

Comment:
could possibly attach this to the function to keep namespacing similar to what we had

```
# mod
entrypoint_waiter.Timeout = EntrypointWaiterTimeout

# usage
try:
    with entrypoint_waiter(foo, 'bar')`
except entrypoint_waiter.Timeout:
    pass
```

not sure what i think about this. though otherwise one might wonder why this exception doesn't live in `nameko.exceptions` with everything else